### PR TITLE
Add shipping example

### PR DESCRIPTION
**Pull request by marcusolsson**
Adding a new example based on the [goddd](https://github.com/marcusolsson/goddd) project.

### CODE REVIEW

**Comment by marcusolsson** [comment]
There still is some cleaning up to do. The application services are using both DTOs and domain objects. Will fix when I manage to find the time to do it. I think it's still a good example though.

**Comment by marcusolsson** [comment]
Created an issue https://github.com/marcusolsson/goddd/issues/12

**Comment by peterbourgon** [comment]
Any insight why [this is crashing during test](https://travis-ci.org/go-kit/kit/jobs/115120992)?

**Comment by marcusolsson** [comment]
Probably because of [this](https://travis-ci.org/go-kit/kit/jobs/114921125)

**Comment by peterbourgon** [comment]
Yeah I see what's happening, but I don't know where or why. You think it's just some bug in tip at the moment?

**Comment by marcusolsson** [comment]
That would be my guess. Seems to be some issue with the linker?

**Comment by marcusolsson** [comment]
Maybe [this](https://github.com/golang/go/commit/862b9ddda7d5163926ca0ab20f23f261a70cfe80). Will continue looking.

**Comment by marcusolsson** [comment]
On second thought, that would've made the previous build fail as well.

**Comment by basvanbeek** [comment]
if you scroll down (it's hard to see) you will notice it fails when the kit/log test is run somehow

**Comment by marcusolsson** [comment]
Yeah, `master` and all the pull requests are currently failing when building using go:tip. Judging by the stacktrace it seems to be related to the Go linker, although the problem may very well be elsewhere.

**Comment by peterbourgon** [comment]
https://github.com/golang/go/issues/14783

**Comment by peterbourgon** [comment]
I've taken tip out of the Travis buildset, can you rebase on master please?

**Comment by marcusolsson** [comment]
Rebased and squashed.

File: examples/shipping/README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+shipping
+========
```
Comment (peterbourgon) [inline examples/shipping/README.md]: Please use consistent hash-prefixed headers throughout.

File: examples/shipping/booking/endpoint.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
+package booking
+
+import (
+	"time"
+
+	"github.com/go-kit/kit/endpoint"
+	"github.com/go-kit/kit/examples/shipping/cargo"
+	"github.com/go-kit/kit/examples/shipping/location"
+	"golang.org/x/net/context"
```
Comment (peterbourgon) [inline examples/shipping/booking/endpoint.go]: Please move third-party imports to their own whitespace-separated block, after stdlib and above go-kit. (Further comments of this type elided.)

File: examples/shipping/booking/logging.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
+package booking
+
+import (
+	"time"
+
+	"github.com/go-kit/kit/log"
+
```
Comment (peterbourgon) [inline examples/shipping/booking/logging.go]: Can remove newline here.

File: examples/shipping/booking/transport.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+	"net/http"
+	"time"
+
+	kitlog "github.com/go-kit/kit/log"
+	kithttp "github.com/go-kit/kit/transport/http"
+	"golang.org/x/net/context"
+
+	"github.com/gorilla/mux"
+	"github.com/go-kit/kit/examples/shipping/cargo"
+	"github.com/go-kit/kit/examples/shipping/location"
```
Comment (peterbourgon) [inline examples/shipping/booking/transport.go]: Please group as

```go
import (
    "stdlib"
    
    "thirdparty"
    
    "go-kit"
)
```

File: examples/shipping/handling/logging.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	"github.com/go-kit/kit/examples/shipping/location"
+	"github.com/go-kit/kit/examples/shipping/voyage"
+)
+
+type loggingService struct {
+	logger log.Logger
+	Service
+}
+
+// NewLoggingService ...
```
Comment (peterbourgon) [inline examples/shipping/handling/logging.go]: Comment?

File: examples/shipping/handling/transport.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	"golang.org/x/net/context"
+
+	kitlog "github.com/go-kit/kit/log"
+	kithttp "github.com/go-kit/kit/transport/http"
+
+	"github.com/gorilla/mux"
+	"github.com/go-kit/kit/examples/shipping/cargo"
+	"github.com/go-kit/kit/examples/shipping/location"
+	"github.com/go-kit/kit/examples/shipping/voyage"
+)
```
Comment (peterbourgon) [inline examples/shipping/handling/transport.go]: Import grouping.

File: examples/shipping/handling/transport.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+		makeRegisterIncidentEndpoint(hs),
+		decodeRegisterIncidentRequest,
+		encodeResponse,
+		opts...,
+	)
+
+	r.Handle("/handling/v1/incidents", registerIncidentHandler).Methods("POST")
+
+	return r
+}
```
Comment (peterbourgon) [inline examples/shipping/handling/transport.go]: I appreciate the symmetry of all of these methods.

File: examples/shipping/handling/transport.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+		w.WriteHeader(http.StatusOK)
+	case cargo.ErrUnknown:
+		w.WriteHeader(http.StatusNotFound)
+	case ErrInvalidArgument:
+		w.WriteHeader(http.StatusBadRequest)
+	default:
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+	json.NewEncoder(w).Encode(map[string]interface{}{
+		"error": err.Error(),
```
Comment (peterbourgon) [inline examples/shipping/handling/transport.go]: Will this panic if err == nil?

File: examples/shipping/routing/proxying.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+	"github.com/go-kit/kit/circuitbreaker"
+	"github.com/go-kit/kit/endpoint"
+	"github.com/go-kit/kit/examples/shipping/cargo"
+	"github.com/go-kit/kit/examples/shipping/location"
+	"github.com/go-kit/kit/examples/shipping/voyage"
+	"golang.org/x/net/context"
+
+	kithttp "github.com/go-kit/kit/transport/http"
+)
```
Comment (peterbourgon) [inline examples/shipping/routing/proxying.go]: Import grouping.

File: examples/shipping/routing/proxying.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+			})
+		}
+
+		itineraries = append(itineraries, cargo.Itinerary{Legs: legs})
+	}
+
+	return itineraries
+}
+
+// ServiceMiddleware ...
```
Comment (peterbourgon) [inline examples/shipping/routing/proxying.go]: Comment.

File: examples/shipping/routing/proxying.go [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+type ServiceMiddleware func(Service) Service
+
+// NewProxyingMiddleware ...
```
Comment (peterbourgon) [inline examples/shipping/routing/proxying.go]: Comment.

File: examples/shipping/tracking/logging.go [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+	"github.com/go-kit/kit/log"
+)
+
+type loggingService struct {
+	logger log.Logger
+	Service
+}
+
+// NewLoggingService ...
```
Comment (peterbourgon) [inline examples/shipping/tracking/logging.go]: Comment.

File: examples/shipping/tracking/transport.go [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+	"errors"
+	"net/http"
+
+	"golang.org/x/net/context"
+
+	kitlog "github.com/go-kit/kit/log"
+	kithttp "github.com/go-kit/kit/transport/http"
+	"github.com/gorilla/mux"
+	"github.com/go-kit/kit/examples/shipping/cargo"
+)
```
Comment (peterbourgon) [inline examples/shipping/tracking/transport.go]: Import grouping.

File: examples/shipping/tracking/transport.go [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+	case cargo.ErrUnknown:
+		w.WriteHeader(http.StatusNotFound)
+	case ErrInvalidArgument:
+		w.WriteHeader(http.StatusBadRequest)
+	default:
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	json.NewEncoder(w).Encode(map[string]interface{}{
+		"error": err.Error(),
```
Comment (peterbourgon) [inline examples/shipping/tracking/transport.go]: Likewise what if err == nil?

File: examples/shipping/booking/transport.go [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+		w.WriteHeader(http.StatusOK)
+	case cargo.ErrUnknown:
+		w.WriteHeader(http.StatusNotFound)
+	case ErrInvalidArgument:
+		w.WriteHeader(http.StatusBadRequest)
+	default:
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+	json.NewEncoder(w).Encode(map[string]interface{}{
+		"error": err.Error(),
```
Comment (peterbourgon) [inline examples/shipping/booking/transport.go]: Likewise what if err == nil?

File: examples/shipping/booking/transport.go [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+	case nil:
+		w.WriteHeader(http.StatusOK)
+	case cargo.ErrUnknown:
+		w.WriteHeader(http.StatusNotFound)
+	case ErrInvalidArgument:
+		w.WriteHeader(http.StatusBadRequest)
+	default:
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+	json.NewEncoder(w).Encode(map[string]interface{}{
```
Comment (peterbourgon) [inline examples/shipping/booking/transport.go]: Missing Content-Type?

File: examples/shipping/handling/transport.go [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+	case nil:
+		w.WriteHeader(http.StatusOK)
+	case cargo.ErrUnknown:
+		w.WriteHeader(http.StatusNotFound)
+	case ErrInvalidArgument:
+		w.WriteHeader(http.StatusBadRequest)
+	default:
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+	json.NewEncoder(w).Encode(map[string]interface{}{
```
Comment (peterbourgon) [inline examples/shipping/handling/transport.go]: Missing Content-Type?

**Comment by peterbourgon** [comment]
This is so bloody elegant. I'm so happy. :)

**Comment by marcusolsson** [comment]
I'm the [happy one](https://twitter.com/enneff/status/705007880638844928) here :) About to sign off, fixing comments first thing tomorrow!

**Comment by marcusolsson** [comment]
Really appreciate you taking the time reviewing this. I've left the commits unsquashed for reviewing. As a sidenote, I believe the comment regarding panic on `err == nil` applies to [this example](https://github.com/go-kit/kit/blob/master/examples/profilesvc/transport.go#L250-L251) as well. I'm thinking that you will not/should not be even be calling the encodeError on `err == nil` so I just removed the case to make that clearer.

**Comment by peterbourgon** [comment]
Right you are, thanks for spotting that, and fixed!

**Pull request merged**